### DISCUSSION TITLE
How does zagjs's machine differ with xstate's machine ?

### DISCUSSION THREAD

**Discussion by astahmer** [chakra-ui/zag#355] [General]
I think it would be nice to have an exhaustive comparison page in the documentation site with a list of differences, this might also help if you intend to contribute to `@xstate/fsm`, or [to keep in sync with `xstate` futures updates](https://github.com/chakra-ui/zag/issues/56)

From what I found looking quickly at the source code: 
- the core package use valtio inlined on the `state` to make it mutable (which means the `context` also is mutable !, this probably make `actions` easier to write)
- introduce is a `computed` context ! [looks like this abandoned RFC](https://github.com/statelyai/rfcs/pull/13)
- introduce a [watch](https://github.com/chakra-ui/zag/blob/d0484fcbea6f7a4cafa3aa9ec66415d8b841300c/packages/core/src/types.ts#L336) property which invokes callbacks on context/computedContext changes
- [`Guard combinations (and, or, not)`](https://github.com/chakra-ui/zag/issues/56) as mentioned here, that looks SO good
- Activities, that have been [since deprecated in xstate's machine](https://xstate.js.org/docs/guides/activities.html#activities)
- introduce [`every`](https://github.com/chakra-ui/zag/blob/d0484fcbea6f7a4cafa3aa9ec66415d8b841300c/packages/core/src/types.ts#L224) that looks like `after` but using a repeating interval
- renamed [`cond`](https://xstate.js.org/docs/guides/guards.html#serializing-guards) to `guard` 
- the react package use valtio/react inlined for useMachine/useActor

(and it looks like valtio usage makes everything simpler 😅 )

**Comment by anubra266** [comment]
@astahmer At the time zag started, there were lots of parts of xstate that were not needed for zag, would have been a lot of unnecessary overhead for the zag ecosystem. @segunadebayo would be better suited to address this though.

**Comment by astahmer** [comment]
btw, is there a way to convert a @zag-js machine to an xstate machine ?
cause if not that means using zag components & my own xstate machines in the same project duplicates the js downloaded which is unfortunate for such related behavior 😢 

EDIT: for the context, I just noticed there is no `invoke` property on state nodes, which I find a bit awkward ? how would you spawn another machine or listen to a callback/promise using @zag-js's createMachine ?

**Reply by anubra266** [reply]
@astahmer you can spawn `machines` in zag. An example is the [toast machine](https://github.com/chakra-ui/zag/blob/86e3dabf65a634e15cb38f656d2f493129b3ee10/packages/machines/toast/src/toast-group.machine.ts#L64)

**Comment by astahmer** [comment]
also, can you invoke a @zag-js machine from an xstate machine ? or is the compatibility mostly limited to/for the visualizer ?

**Reply by anubra266** [reply]
@astahmer You can't do that. The visualizer uses xstate, we [convert](https://github.com/chakra-ui/zag/blob/main/scripts/visualize.ts) the zag syntax to xstate syntax, then xstate uses [that](https://github.com/chakra-ui/zag/tree/main/.xstate) in our visualizer.
